docs/content/specs/terraform/_index.md

```diff
@@ -533,6 +543,466 @@ Bad examples:
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR16 - Category: Code Style - Name of a `variable` **MUST** follow rules
+
+The naming of a `variable` should follow [HashiCorp's naming rule](https://www.terraform.io/docs/extend/best-practices/naming.html).
+
+`variable` used as feature switches should use positive statement, use `xxx_enabled` instead of `xxx_disabled`. Avoid double negatives like `!xxx_disabled`.
+
+Please use `xxx_enabled` instead of `xxx_disabled` as name of a `variable`.
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR17 - Category: Code Style - Every `variable` **MUST** come with a `description`
+
+The target audience of `description` is the module users.
+
+For a newly created `variable` (Eg. `variable` for switching `dynamic` block on-off), it's `description` should precisely describe the input parameter's purpose and the expected data type. `description` should not contain any information for module developers, this kind of information can only exist in code comments.
+
+For `object` type `variable`, `description` can be composed in HEREDOC format:
+
+```terraform
+variable "kubernetes_cluster_key_management_service" {
+  type = object({
+    key_vault_key_id         = string
+    key_vault_network_access = optional(string)
+  })
+  default     = null
+  description = <<-EOT
+  - `key_vault_key_id` - (Required) Identifier of Azure Key Vault key. See [key identifier format](https://learn.microsoft.com/en-us/azure/key-vault/general/about-keys-secrets-certificates#vault-name-and-object-name) for more details. When Azure Key Vault key management service is enabled, this field is required and must be a valid key identifier. When `enabled` is `false`, leave the field empty.
+  - `key_vault_network_access` - (Optional) Network access of the key vault Network access of key vault. The possible values are `Public` and `Private`. `Public` means the key vault allows public access from all networks. `Private` means the key vault disables public access and enables private link. Defaults to `Public`.
+EOT
+}
+```
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR18 - Category: Code Style - Every `variable` **MUST** have an appropriate `type`
+
+`type` **MUST** be defined for every `variable`. `type` should be as precise as possible, `any` can only be defined with adequate reasons.
+
+* Use `bool` instead of `string` or `number` for `true/false`
+* Use `string` for text
+* Use concrete `object` instead of `map(any)`
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR19 - Category: Code Style - `variable` containing confidential data should be declared as `sensitive = true`
+
+If `variable`'s `type` is `object` and contains one or more fields that would be assigned to a `sensitive` argument, then this whole `variable` should be declared as `sensitive = true`, otherwise you should extract sensitive field into separated variable block with `senstive = true`.
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR20 - Category: Code Style - Declare `nullable = false` when it's possible
```

Comment:
Should we state that for simple values, e.g. string, number. Then nullable is ok. Null provides an explicit assertion that the value is not used. Whereas empty string and number values are ambiguous.

However for collection values, e.g. list, set, map. Here nullable should be false and we should use the `[]`, `[]`, `{}` notation respectively?